pynativelib-proposal.rst

```diff
@@ -0,0 +1,797 @@
+accepted.)
+
+So in a ``pkg-config`` file, you just write down all your compile
+arguments and linker arguments (they call them ``cflags`` and
+``libs``) using GCC style, and ``pkg-config`` knows about ``-I`` and
+``-L`` and ``-l``, and if you really want just those parts of the
+command line it can pull them out, and if you want MSVC-style syntax
+then you pass ``--msvc`` and it will translate ``-L`` and ``-l`` for
+you according to the above table.
+
+(``pkg-config`` also has some special casing for ``-rpath``, but
+``-rpath`` is never useful for us, because we never know where the
+pynativelib library will be installed on the user filesystem. AFAICT
+these are the only cases that ``pkg-config`` thinks need special
+casing.)
+
+What I like about the ``pkg-config`` approach is that as a user you
+basically just say "give me {GCC, MSVC} {compile, link} flags", and
+then the gory details are encapsulated. This is strictly more powerful
+than the distutils approach. Consider the case of a library where if
+you use it with GCC you need to pass ``--pthread`` as part of your
+link flags -- but with MSVC you don't want to pass this. In the
+distutils approach there's simply now way to do this.
+
+Therefore, we adopt a more ``pkg-config``-style approach, plus a
+convenience adaptor to the legacy distutils-style approach.
+
+- ``compile_args(compiler_style)``, ``link_args(compiler_style)``:
+  Mandatory argument ``compiler_style`` should be ``"gcc"`` or
+  ``"msvc"`` (or potentially other styles later if it turns out to be
+  relevant). Returns a list of strings like ``["-Ifoo", "-DFOO"]``.
+
+  Note: Similarly to ``enable``, if successfully linking to
+  pynativelib1 requires some arguments be added to also link to
+  pynativelib2, then these functions should call their counterparts in
+  pynativelib2 and merge them into the results.
+
+- ``distutils_extension_kwargs()``: returns a dict of kwargs to pass
+  to `the distutils ``Extension`` constructor
+  <<https://docs.python.org/2/distutils/apiref.html#distutils.core.Extension>`_,
+  or raises an error if this is impossible (e.g. if there are
+  necessary arguments that cannot be translated into the distutils
+  interface).
+
+  This should only be used as a workaround for build systems where we
+  can't use the ``{compile,link}_args`` API.
+
+``pkg-config`` also provides some options for pulling out particular
+parts of the args (e.g. ``--cflags-only-I`` + ``--cflags-only-other``
+= ``--cflags``). I'm not sure why this is needed. We can add it later
+if it turns out to be necessary.
+
+FIXME: in some cases we might not be able to convince some code to
+link against a specific library just by controlling the arguments
+passed to the compiler; for example, there's no argument you can pass
+to ``gfortran`` to tell it "please link against
+``pynativelib_libgfortran__libgfortran`` instead of
+``libgfortran``. Maybe we need some additional function in the
```

Comment:
Don't quote me on this until we try it, but I think with gfortran and g++ if you build just `.o` object files using `-c` with the fortran/c++ specific driver, but then link with gcc, I think you should be able to add your own custom linking to a different libgfortran or libstdc++. Doing the same with libgcc_s, if ever necessary, requires dropping down to linking with `ld` I think which can get messier.